pdns/protobuf-dnstap.cc

```diff
@@ -0,0 +1,248 @@
+    message->set_response_port(ntohs(responder.sin4.sin_port));
+  }
+#endif /* HAVE_PROTOBUF */
+}
+
+void DnstapProtoBufMessage::setResponder(const std::string& responder)
+{
+#ifdef HAVE_PROTOBUF
+  dnstap::Message* message = proto_message.mutable_message();
+  message->set_response_address(responder);
+#endif /* HAVE_PROTOBUF */
+}
+
+void DnstapProtoBufMessage::setResponseCode(uint8_t rcode)
+{
+  // There is no specific response code field to set for dnstap
+  // messages. However, it can be found in the response_message.
+}
+
+void DnstapProtoBufMessage::setTime(time_t sec, uint32_t usec)
+{
+#ifdef HAVE_PROTOBUF
+  const dnstap::Message message = proto_message.message();
+  if (message.type() == dnstap::Message_Type_FORWARDER_QUERY) {
+    setQueryTime(sec, usec);
+  } else if (message.type() == dnstap::Message_Type_FORWARDER_RESPONSE) {
+    setResponseTime(sec, usec);
+  } else {
+    warnlog("TODO");
+  }
+#endif /* HAVE_PROTOBUF */
+}
+
+void DnstapProtoBufMessage::setType(DNSProtoBufMessage::DNSProtoBufMessageType type)
+{
+#ifdef HAVE_PROTOBUF
+  // Required. Only possible value.
+  proto_message.set_type(dnstap::Dnstap_Type_MESSAGE);
+
+  dnstap::Message* message = proto_message.mutable_message();
+  switch(type) {
+  case DNSProtoBufMessage::DNSProtoBufMessageType::Query:
+    message->set_type(dnstap::Message_Type_FORWARDER_QUERY);
+    break;
+  case DNSProtoBufMessage::DNSProtoBufMessageType::Response:
+    message->set_type(dnstap::Message_Type_FORWARDER_RESPONSE);
+    break;
+  case DNSProtoBufMessage::DNSProtoBufMessageType::OutgoingQuery:
+    message->set_type(dnstap::Message_Type_RESOLVER_QUERY);
+    break;
+  case DNSProtoBufMessage::DNSProtoBufMessageType::IncomingResponse:
+    message->set_type(dnstap::Message_Type_RESOLVER_RESPONSE);
+    break;
+  default:
+    throw std::runtime_error("Unsupported protobuf type: "+std::to_string(type));
+  }
+#endif /* HAVE_PROTOBUF */
+}
+
+std::string DnstapProtoBufMessage::toDebugString() const
+{
+#ifdef HAVE_PROTOBUF
+  return proto_message.DebugString();
+#else
+  return std::string();
+#endif /* HAVE_PROTOBUF */
+}
+
+#ifdef HAVE_PROTOBUF
+
+void DnstapProtoBufMessage::setInitialRequestID(const boost::uuids::uuid& uuid)
+{
+  std::string* identity = proto_message.mutable_identity();
+  identity->resize(uuid.size());
+  std::copy(uuid.begin(), uuid.end(), identity->begin());
+}
+
+void DnstapProtoBufMessage::setUUID(const boost::uuids::uuid& uuid)
+{
+  std::string* extra = proto_message.mutable_extra();
+  extra->resize(uuid.size());
+  std::copy(uuid.begin(), uuid.end(), extra->begin());
+}
+
+void DnstapProtoBufMessage::update(const boost::uuids::uuid& uuid, const ComboAddress* requestor, const ComboAddress* responder, bool isTCP, uint16_t id)
+{
+  struct timespec ts;
+  gettime(&ts, true);
+  setTime(ts.tv_sec, ts.tv_nsec / 1000);
+  setUUID(uuid);
+
+  char identity[MAXHOSTNAMELEN+1];
+  if (gethostname(identity, MAXHOSTNAMELEN) == 0) {
```

Comment:
Please don't call `gethostname()` for every update() call. Either pass the host name to the constructor or compute it only once there and store it. You might even want to allow setting the identity as a parameter to `newFrameStreamLogger()`.